File: 2022/day1-calorieCounting.js
const { readFile } = require('../utils');

(async () => {
  const input = await readFile('day1.txt', Number);
  console.log(calorieCounting(input));
  console.log(calorieCountingPart2(input));
})();

const calorieCounting = (input) => {
  let caloriesByElf = 0;
  return input.reduce((result, elem) => {
    if (!elem) {
      const maxCaloriesByElf = caloriesByElf;
      caloriesByElf = 0;
      if (maxCaloriesByElf > result) {
        return maxCaloriesByElf;
      }
      return result;
    }
    caloriesByElf += elem;
    return result;
  }, 0);
};

const calorieCountingPart2 = (input) => {
  let caloriesByElf = 0;
  const caloriesByElfList = input.reduce((result, elem) => {
    if (!elem) {
      const newResult = [...result, caloriesByElf];
      caloriesByElf = 0;
      return newResult;
    }
    caloriesByElf += elem;
    return result;
  }, []);
  caloriesByElfList.sort((a, b) => b - a);
  return caloriesByElfList[0] + caloriesByElfList[1] + caloriesByElfList[2];
};

File: 2022/Kotlin/src/main/kotlin/day1-calorieCounting.kt
fun main() {
  val input: Array<Int> =
    readInput("day1.txt") { if (it == "") 0 else it.toInt() }.toArray(arrayOf<Int>())
  println(calorieCounting(input))
  println(calorieCountingPart2(input))
}

fun calorieCounting(input: Array<Int>): Int {
  var caloriesByElf = 0
  return input.fold(0) { result, elem ->
    if (elem == 0) {
      val maxCaloriesByElf = caloriesByElf
      caloriesByElf = 0
      if (maxCaloriesByElf > result) {
        return@fold maxCaloriesByElf
      }
      return@fold result
    }
    caloriesByElf += elem
    return@fold result
  }
}

fun calorieCountingPart2(input: Array<Int>): Int {
  var caloriesByElf = 0
  val caloriesByElfList =
    input.fold(arrayOf<Int>()) { result, elem ->
      if (elem == 0) {
        val newResult = result + caloriesByElf
        caloriesByElf = 0
        return@fold newResult
      }
      caloriesByElf += elem
      return@fold result
    }
  caloriesByElfList.sortDescending()
  return caloriesByElfList[0] + caloriesByElfList[1] + caloriesByElfList[2]
}
